File: ruby/examples/webhooks_e2e.rb
require "json"
require_relative "../lib/quicknode_sdk"

qn = QuicknodeSdk::SDK.from_env

# list_webhooks is arity-1 native: exercise bare, kwargs, and positional hash.
before = qn.webhooks.list_webhooks({})
puts "webhooks before: #{before[:data].length} (total=#{before.dig(:pageInfo, :total)})"
raise "list_webhooks bare broke" unless qn.webhooks.list_webhooks.keys.sort == before.keys.sort
raise "list_webhooks kwargs splat broke" unless qn.webhooks.list_webhooks(**{}).keys.sort == before.keys.sort

# get_enabled_count is arity-0 native: exercise bare, positional empty hash,
# and kwargs splat — all three must reach the no-args branch of the
# dispatcher cleanly.
count = qn.webhooks.get_enabled_count
puts "enabled count: #{count[:total]}"
raise "get_enabled_count positional hash broke" unless qn.webhooks.get_enabled_count({}).keys.sort == count.keys.sort
raise "get_enabled_count kwargs splat broke" unless qn.webhooks.get_enabled_count(**{}).keys.sort == count.keys.sort

destination_attributes = JSON.generate({
  url: "https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef",
  compression: "none"
})

template_args = JSON.generate({
  templateId: "evmWalletFilter",
  templateArgs: { wallets: ["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"] }
})

webhook = qn.webhooks.create_webhook_from_template(
  name: "E2E Test Webhook",
  network: "ethereum-mainnet",
  destination_attributes_json: destination_attributes,
  template_args_json: template_args
)
id = webhook[:id]
puts "created: #{id} | #{webhook[:status]}"

fetched = qn.webhooks.get_webhook(id: id)
puts "fetched: #{fetched[:id]} | #{fetched[:name]}"

updated = qn.webhooks.update_webhook(id: id, name: "E2E Test Webhook Updated")
puts "updated name: #{updated[:name]}"

qn.webhooks.pause_webhook(id: id)
puts "paused"

qn.webhooks.activate_webhook(id: id, start_from: "latest")
puts "activated"

qn.webhooks.delete_webhook(id: id)
puts "deleted: #{id}"
sleep 1

after = qn.webhooks.list_webhooks({})
puts "webhooks after: #{after[:data].length} (total=#{after.dig(:pageInfo, :total)})"

File: python/examples/webhooks_e2e.py
import asyncio

from quicknode_sdk import (
    EvmContractEventsArgs,
    EvmContractEventsTemplate,
    EvmWalletFilterArgs,
    EvmWalletFilterTemplate,
    QuicknodeSdk,
    WebhookDestinationAttributes,
)


async def main():
    qn = QuicknodeSdk.from_env()

    before = await qn.webhooks.list_webhooks()
    print(f"webhooks before: {len(before.data)} (total={before.page_info.total})")

    count = await qn.webhooks.get_enabled_count()
    print(f"enabled count: {count.total}")

    webhook = await qn.webhooks.create_webhook_from_template(
        name="E2E Test Webhook",
        network="ethereum-mainnet",
        destination_attributes=WebhookDestinationAttributes(
            url="https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef",
            compression="none",
        ),
        template_args=EvmWalletFilterArgs(
            EvmWalletFilterTemplate(wallets=["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"])
        ),
    )
    webhook_id = webhook.id
    print(f"created: {webhook_id} | {webhook.status}")

    fetched = await qn.webhooks.get_webhook(webhook_id)
    print(f"fetched: {fetched.id} | {fetched.name}")

    updated = await qn.webhooks.update_webhook(webhook_id, name="E2E Test Webhook Updated")
    print(f"updated name: {updated.name}")

    await qn.webhooks.pause_webhook(webhook_id)
    print("paused")

    await qn.webhooks.activate_webhook(webhook_id, start_from="latest")
    print("activated")

    await qn.webhooks.delete_webhook(webhook_id)
    print(f"deleted: {webhook_id}")

    # Exercise the evm-contract-events template, which carries the multi-word
    # event_hashes field. The API expects eventHashes on the wire.
    ce_webhook = await qn.webhooks.create_webhook_from_template(
        name="E2E Test Webhook (evmContractEvents)",
        network="ethereum-mainnet",
        destination_attributes=WebhookDestinationAttributes(
            url="https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef",
            compression="none",
        ),
        template_args=EvmContractEventsArgs(
            EvmContractEventsTemplate(
                contracts=["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"],
                event_hashes=[
                    "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"
                ],
            )
        ),
    )
    print(f"created (evmContractEvents): {ce_webhook.id} | {ce_webhook.status}")
    await qn.webhooks.delete_webhook(ce_webhook.id)
    print(f"deleted (evmContractEvents): {ce_webhook.id}")

    after = await qn.webhooks.list_webhooks()
    print(f"webhooks after: {len(after.data)} (total={after.page_info.total})")


asyncio.run(main())
